File: backend/app/models/tender.py
"""
Tender AI Platform - Database Models
Matches TypeScript types exactly
"""

from sqlalchemy import (
    Column, String, Text, DateTime, Enum, JSON, ForeignKey, Integer, Boolean
)
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.core.database import Base
import uuid
import enum


class TenderStatus(str, enum.Enum):
    PENDING = "PENDING"
    LISTED = "LISTED"
    ANALYZED = "ANALYZED"
    ERROR = "ERROR"


class TenderType(str, enum.Enum):
    AOON = "AOON"
    AOOI = "AOOI"


class DocumentType(str, enum.Enum):
    AVIS = "AVIS"
    RC = "RC"
    CPS = "CPS"
    ANNEXE = "ANNEXE"
    UNKNOWN = "UNKNOWN"


class ExtractionMethod(str, enum.Enum):
    DIGITAL = "DIGITAL"
    OCR = "OCR"


class Tender(Base):
    """Main tender record"""
    __tablename__ = "tenders"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    external_reference = Column(String(255), index=True)
    source_url = Column(Text, nullable=False)
    status = Column(
        Enum(TenderStatus), 
        default=TenderStatus.PENDING, 
        nullable=False,
        index=True
    )
    
    # Scraping metadata
    scraped_at = Column(DateTime(timezone=True), server_default=func.now())
    download_date = Column(String(10))  # YYYY-MM-DD format
    
    # AI-extracted metadata (Phase 1 - Avis)
    # Stored as JSONB for flexibility with provenance tracking
    avis_metadata = Column(JSONB, nullable=True)
    
    # AI-extracted metadata (Phase 2 - Universal deep analysis)
    universal_metadata = Column(JSONB, nullable=True)
    
    # Error tracking
    error_message = Column(Text, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # Relationships
    documents = relationship("TenderDocument", back_populates="tender", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Tender {self.external_reference or self.id}>"


class TenderDocument(Base):
    """Document extracted from tender ZIP"""
    __tablename__ = "tender_documents"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tender_id = Column(UUID(as_uuid=True), ForeignKey("tenders.id", ondelete="CASCADE"), nullable=False)
    
    document_type = Column(Enum(DocumentType), default=DocumentType.UNKNOWN)
    filename = Column(String(500), nullable=False)
    
    # Extracted text content
    raw_text = Column(Text, nullable=True)
    page_count = Column(Integer, nullable=True)
    
    # Extraction method used
    extraction_method = Column(Enum(ExtractionMethod), nullable=True)
    
    # File metadata (stored in memory, not on disk)
    file_size_bytes = Column(Integer, nullable=True)
    mime_type = Column(String(100), nullable=True)
    
    # Timestamps
    extracted_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    tender = relationship("Tender", back_populates="documents")

    def __repr__(self):
        return f"<TenderDocument {self.filename}>"


class ScraperJob(Base):
    """Track scraper execution history"""
    __tablename__ = "scraper_jobs"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    
    target_date = Column(String(50), nullable=False)  # Date range: "YYYY-MM-DD to YYYY-MM-DD"
    status = Column(String(50), default="RUNNING")  # RUNNING, COMPLETED, FAILED, STOPPED
    
    # Stats
    total_found = Column(Integer, default=0)
    downloaded = Column(Integer, default=0)
    failed = Column(Integer, default=0)
    extracted = Column(Integer, default=0)
    
    # Timing
    started_at = Column(DateTime(timezone=True), server_default=func.now())
    completed_at = Column(DateTime(timezone=True), nullable=True)
    elapsed_seconds = Column(Integer, nullable=True)
    
    # Current phase for progress tracking
    current_phase = Column(String(100), default="Initializing")
    
    # Error log
    error_log = Column(Text, nullable=True)

    def __repr__(self):
        return f"<ScraperJob {self.target_date} - {self.status}>"

File: src/types/tender.ts
// Tender AI Platform — Type Definitions (V1)
// Authoritative schemas matching backend PostgreSQL

export type TenderStatus = 'PENDING' | 'LISTED' | 'ANALYZED' | 'ERROR';
export type TenderType = string | null;  // Flexible: AOON, AOOI, AOO, AOR, Consultation, etc.
export type DocumentType = 'AVIS' | 'RC' | 'CPS' | 'ANNEXE' | 'UNKNOWN';

// Provenance tracking for every extracted field
export interface TrackedValue<T> {
  value: T | null;
  source_document: DocumentType | 'WEBSITE' | null;
  source_date: string | null;
}

// Lot structure
export interface TenderLot {
  lot_number: string | null;
  lot_subject: string | null;
  lot_estimated_value: string | null;
  caution_provisoire: string | null;
  // Deep analysis fields (Phase 2)
  caution_definitive_percentage?: string | null;
  estimated_caution_definitive_value?: string | null;
  execution_date?: string | null;
  items?: TenderItem[];
}

// Item within a lot (Deep analysis)
export interface TenderItem {
  item_name: string | null;
  quantity: string | null;
  technical_description_full: string | null;
}

// Multilingual keywords
export interface TenderKeywords {
  keywords_fr: string[];
  keywords_eng: string[];
  keywords_ar: string[];
}

// Website-scraped extended metadata (authoritative)
export interface WebsiteExtendedMetadata {
  acheteur_public: TrackedValue<string>;  // Buyer/purchasing entity
  lieu_execution: TrackedValue<string>;   // Execution location
  estimation_ttc: TrackedValue<string>;   // Estimated value DHS TTC
  lieu_ouverture_plis: TrackedValue<string>; // Bid opening location
  caution_provisoire_website: TrackedValue<string>; // Provisional guarantee from website
  contact_administratif: TrackedValue<string>; // Raw contact info to be structured by AI
}

// Avis metadata schema (Phase 1 - Night Shift)
export interface AvisMetadata {
  reference_tender: TrackedValue<string>;
  tender_type: TrackedValue<TenderType>;
  issuing_institution: TrackedValue<string>;
  // NEW: Lieu d'exécution (website + docs fallbacks)
  execution_location?: TrackedValue<string>;
  submission_deadline: {
    date: TrackedValue<string>;
    time: TrackedValue<string>;
  };
  folder_opening_location: TrackedValue<string>;
  subject: TrackedValue<string>;
  total_estimated_value: TrackedValue<string> & { currency?: string | null };
  lots: TenderLot[];
  keywords: TenderKeywords;
  // Website extended metadata (authoritative)
  website_extended?: WebsiteExtendedMetadata;
}

// Deep analysis data for a lot (complementary to AvisMetadata)
export interface LotDeepData {
  lot_number: string;
  caution_definitive_percentage?: TrackedValue<string>;
  estimated_caution_definitive_value?: {
    value: string | null;
    currency: string | null;
  };
  execution_delay?: {
    value: string | null;
    unit: string | null;
    source_document: DocumentType | null;
  };
  items: TenderItem[];
}

// Structured contact info (AI-parsed from raw contact text)
export interface StructuredContact {
  name: string | null;
  role: string | null;
  phone: string | null;
  email: string | null;
  address: string | null;
}

// Additional conditions from deep analysis
export interface AdditionalConditions {
  qualification_criteria: string | null;
  required_documents: string[];
  warranty_period: string | null;
  payment_terms: string | null;
}

// Universal fields schema (Phase 2 - User Shift)
// Contains ONLY complementary data not in AvisMetadata

export interface UniversalMetadata {
  institution_address: TrackedValue<string>;
  lots_deep_data: LotDeepData[];
  additional_conditions: AdditionalConditions;
  // AI-structured contact from website raw text
  structured_contact?: StructuredContact;
}

// Document extracted from tender
export interface TenderDocument {
  id: string;
  tender_id: string;
  document_type: DocumentType;
  filename: string;
  raw_text: string | null;
  page_count: number | null;
  extraction_method: 'DIGITAL' | 'OCR';
  extracted_at: string;
}

// Main Tender record
export interface Tender {
  id: string;
  external_reference: string;
  source_url: string;
  status: TenderStatus;
  
  // Scraped at download time
  scraped_at: string;
  download_date: string;
  
  // Avis metadata (Phase 1)
  avis_metadata: AvisMetadata | null;
  
  // Universal metadata (Phase 2 - on demand)
  universal_metadata: UniversalMetadata | null;
  
  // Related documents
  documents?: TenderDocument[];
  
  // Timestamps
  created_at: string;
  updated_at: string;
}

// API Response types
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

// Scraper log entry
export interface ScraperLogEntry {
  level: 'info' | 'success' | 'warning' | 'error';
  message: string;
  timestamp?: string;
}

// Scraper stats
export interface ScraperStats {
  total: number;
  downloaded: number;
  failed: number;
  elapsed: number;
}

// Scraper status
export interface ScraperStatus {
  is_running: boolean;
  current_phase: string;
  total_tenders: number;
  downloaded: number;
  failed: number;
  elapsed_seconds: number;
  last_run: string | null;
  logs?: ScraperLogEntry[];
  stats?: ScraperStats;
}

// Search/filter params
export interface TenderSearchParams {
  query?: string;
  status?: TenderStatus;
  date_from?: string;
  date_to?: string;
  page?: number;
  per_page?: number;
}
